watertap/tests/test_notebooks.py

```diff
@@ -0,0 +1,98 @@
+"""
+Test Jupyter notebooks in the docs.
+"""
+# stdlib
+import logging
+from pathlib import Path
+import time
+
+# deps
+import nbformat
+from nbconvert.preprocessors import ExecutePreprocessor, CellExecutionError
+import pytest
+
+# package
+from watertap.edb import ElectrolyteDB
+
+# Logging setup
+
+_log = logging.getLogger(__name__)
+
+# Constants
+
+JUPYTER_NB_VERSION = 4
+EXECUTE_TIMEOUT_SEC = 600
+
+# Exceptions
+
+
+class NotebookError(Exception):
+    pass
+
+
+# Utility
+# =======
+
+
+def run_notebook(path: Path):
+    # parse
+    _log.debug(f"begin: parsing '{path}'")
+    with path.open("r", encoding="utf-8") as f:
+        nb = nbformat.read(f, as_version=JUPYTER_NB_VERSION)
+    _log.debug(f"end: parsing '{path}'")
+
+    # execute
+    pp = ExecutePreprocessor(timeout=EXECUTE_TIMEOUT_SEC)
+    _log.debug(f"begin: executing '{path}'")
+    t0 = time.time()
+    try:
+        metadata = {"metadata": {"path": path.parent}}
+        pp.preprocess(nb, metadata)
+    except (CellExecutionError, NameError) as err:
+        _log.warning(f"Failed: {path} :: {err}")
+        raise NotebookError(f"execution error for '{path}': {err}")
+    except TimeoutError as err:
+        _log.warning(f"Timeout: {path} :: {err}")
+        dur, timeout = time.time() - t0, EXECUTE_TIMEOUT_SEC
+        raise NotebookError(f"timeout for '{path}': {dur}s > {timeout}s")
+    dur = time.time() - t0
+    _log.debug(f"end: executing '{path}' duration={dur}s")
+
+
+def find_notebooks(path: Path):
+    return path.glob("**/*.ipynb")
+
+
+@pytest.fixture(scope="session")
+def docs_root(dirname="docs"):
+    p = Path(__file__).parent
+    while not (p / dirname).exists():
+        pp = p.parent
+        if pp == p:
+            raise RuntimeError(f"Could not find '{dirname}' dir")
+        p = pp
+    yield p / dirname
```

Comment:
Yeah skip is a good idea, though that's a bit trickier than desired. Is there just a flag we can use to tell if the package is an archive file instead of a real directory tree?

Reply:
Sorry for the confusion, what I meant with "package mode" is "non-editable installation mode", i.e. where the Python package has its files stored under `.../python3.X/site-packages/watertap/...` (where `docs` is not present) rather than a local clone of the repo (where `docs` is present).

In this case probably checking for `if "site-packages" in Path(__file__).parts` should be enough, although, as I said, this is not specific to the notebook tests and should be implemented as a globally-available utility skipmark/fixture.